Integrate copyright element

bcmn opened the pull request:
Resolves #456 #655

* Integrates copyright component
* Changes fixture data image, to one more typical -- previous had embedded source & logo. [Previous](https://ichef.bbci.co.uk/news/640/cpsprodpb/439A/production/_100960371_syrians_and_asylum_v2-nc.png) | [New](https://ichef.bbci.co.uk/news/640/cpsprodpb/E7DB/production/_101655395_paulineclayton.jpg)
* Changes `test/c0000000002o.json` & `test/c0000000025o.json` to allow for testing of image copyright with & without image caption.
* Tested with Voiceover to ensure AT pauses suitably between the copyright text & a following caption or paragraph, & reads the visually hidden ["image source" text](https://github.com/bbc/simorgh/issues/456#issuecomment-420678019) with a suitable pause.
* Capitalises text in copyright element using CSS, to avoid AT reading it incorrectly
* Transparent background, ensures hidden accessibility full-stop is wholly transparent to avoid showing against the translucent background

- [x] Tests added for new features
- [ ] Test engineer approval

bcmn committed:
Adds example of non-BBC image copyright to test articles

bcmn committed:
Merge branch 'latest' of github.com:bbc/simorgh into integrate_copyright_element

bcmn committed:
Integrates copyright component, changes fixture data to image without embedded source & logo

bcmn committed:
Integrates copyright component, changes fixture data to image without embedded source & logo

bcmn committed:
Merge branch 'latest' of github.com:bbc/simorgh into integrate_copyright_element

bcmn committed:
Merge branch 'integrate_copyright_element' of github.com:bbc/simorgh into integrate_copyright_element

greenc05 commented:
@bcmn lets chat about how we can implement the new accessibility copyright requirements when we know what they are

bcmn committed:
Merge branch 'latest' of github.com:bbc/simorgh into integrate_copyright_element

bcmn committed:
Merge branch 'latest' of github.com:bbc/simorgh into integrate_copyright_element

bcmn committed:
Update integration and code to reflect #653

bcmn committed:
Adds visually hidden text to include "Image source" terminology for AT users

bcmn committed:
Rename copyright div for consistency, runs updateScreenshots

bcmn committed:
Merge branch 'latest' of github.com:bbc/simorgh into integrate_copyright_element

bcmn commented:
Tested with Voiceover this morning to make sure the ["image source" wording](https://github.com/bbc/simorgh/issues/456#issuecomment-420678019) has suitable pauses, etc. 👍

sareh reviewed (dismissed):
Looks great! Thanks for making the changes.

jtart commented on src/app/components/Figure/Copyright/index.jsx:
```diff
@@ -29,4 +32,15 @@ const Copyright = styled.span.attrs({
 
+const Copyright = ({ children }) => (
+  <StyledCopyright>
+    <VisuallyHiddenText>Image source,</VisuallyHiddenText>
```
As we are already passing `children` in, it might be cleaner to pass `VisuallyHiddenText` in as a child. This would allow us to continue to export the styled Copyright component, and would mean that we can handle things like translations (of `Image source`) at a container rather than component level.

What do you think?

bcmn replied:
This is a good point -- from a consistency viewpoint, it seems stuff like ImageCaption hasn't currently split out VisuallyHiddenText to the container like this? It seems like the most similar example I could find in the codebase. 

There's already a ticket for adding the internationalisation to the children, as it's always English (#623) -- perhaps this would fit in there nicely?

jtart replied:
Yep that sounds fine. Let's just do it closer to the time and not let it hold us up now.

bcmn committed:
Changes test article 0000000025 to include two non-BBC copyrightHolders for easy testing

bcmn committed:
Changes prod article 0000000025 to include two non-BBC copyrightHolders for easy testing

jamesbrumpton committed:
Adding Cypress tests for Copyright element

sareh commented on data/test/news/c0000000002o.json:
```diff
@@ -70,7 +78,7 @@
                   "locator": "/cpsprodpb/439A/production/_100960371_syrians_and_asylum_v2-nc.png",
                   "originCode": null,
-                  "copyrightHolder": "BBC"
+                  "copyrightHolder": "Getty Images"
```
I'd suggest making this change on another test article, for example `c0000000005o`, since the images on this article have another logo in the bottom right.
Also, could you please update this table in the Readme, explaining that there are images with Image source attribution on the associated articles: https://github.com/bbc/simorgh/blob/latest/data/README.md

bcmn replied:
Changed to `c0000000005o` & updated readme.

sareh commented on data/prod/news/c0000000025o.json:
```diff
@@ -125,7 +125,7 @@
                   "locator": "/cpsprodpb/A933/production/_101651334_bouquet_pa.jpg",
                   "originCode": null,
-                  "copyrightHolder": "BBC"
+                  "copyrightHolder": "Getty Images"
```
@bcmn The production articles are ones which we will *actually* put in production - they'll be audience-facing. For this reason, we shouldn't change the image attribution to be inaccurate.  

@jimjohnsonrollings Jim - is there an example of an image on the Prod articles that have non-BBC attribution, which we could use for testing?

jimjohnsonrollings replied:
Yes - 25 has both BBC and non-BBC images in.

sareh replied:
Great, we'll add them, then.

bcmn replied:
Updated prod pages to reflect real copyrightHolders & changed cypress tests to reflect.

sareh reviewed (changes requested):
The work on the PR looks great. I am just wary of changing the content of a Prod article in a way that makes it no longer accurate. Have cc'd in Jim to give his thoughts on the matter.

bcmn committed:
Reverts test/c0000000002o data file

bcmn committed:
Changes test/c0000000005o to demo copyright element with and without image caption, updates data readme to reflect

bcmn committed:
Changes prod articles to have accurate copyrightHolder values

bcmn committed:
Updates Cypress tests to reflect accurate prod page copyrightHolders

jtart commented on data/prod/news/c0000000025o.json:
```diff
@@ -443,7 +443,7 @@
                   "locator": "/cpsprodpb/1743E/production/_101649259_850742ef-c5c5-4537-97b1-6d0eec101534.jpg",
                   "originCode": null,
-                  "copyrightHolder": "BBC"
+                  "copyrightHolder": "GETTY IMAGES"
```
Are these meant to be in all caps?

jtart replied:
@jimjohnsonrollings ^

jimjohnsonrollings replied:
They will be mixed case in the data - we should however always display them in upper case.

bcmn committed:
Make test and prod image copyrightHolders match and be correctly cased

bcmn committed:
Figure container capitalises strings passed to Copyright component

bcmn committed:
Make Figure fixture data more representative, update snapshots

bcmn commented:
Fixed test/prod data, Figure container now capitalises the copyright string before passing to Copyright component.

sareh commented on src/app/containers/Figure/index.jsx:
```diff
@@ -1,23 +1,30 @@
 
 const renderCopyright = copyright =>
-  copyright ? <VisuallyHiddenText>{copyright}</VisuallyHiddenText> : null;
+  copyright ? <Copyright>{copyright.toUpperCase()}</Copyright> : null;
```
I recommend using the CSS property `text-transform: uppercase;` in the Copyright component, rather than changing the actual text from mixed cases to upper case. This is so that screenreaders don't read the text as initials, but rather the words.

sareh commented on data/test/news/c0000000005o.json:
```diff
@@ -159,6 +159,78 @@
             ]
           }
         },
+        {
```
Thanks for updating a Test article to capture the two different cases - an image with caption and without a caption, but both having the copyright text.  
Great that you've also updated the Readme file as well. 💯

sareh reviewed (commented):
Great work! You've added test cases to our test data, which is very useful, too. 
I've just got one comment about the implementation of the uppercasing.

bcmn committed:
Switch to using CSS to capitalise Copyright text, instead of altering the string itself

sareh reviewed (dismissed):
Thanks for the changes. Looks great!

BogdanDogaru reviewed (dismissed):
Looks good to me. Nice work!

bcmn committed:
Merge branch 'latest' into integrate_copyright_element

bcmn committed:
Merge branch 'latest' into integrate_copyright_element

bcmn commented:
A few screenshots to document...

Third-party copyright, no caption:
<img width="799" alt="screen shot 2018-09-19 at 10 49 50" src="https://user-images.githubusercontent.com/11646957/45745929-ddf6f400-bbf9-11e8-83de-daba1ce31414.png">

Third-party copyright, with caption:
<img width="795" alt="screen shot 2018-09-19 at 10 49 32" src="https://user-images.githubusercontent.com/11646957/45745981-fcf58600-bbf9-11e8-9053-6621276aba16.png">

First-party copyright, no caption:
<img width="805" alt="screen shot 2018-09-19 at 10 49 59" src="https://user-images.githubusercontent.com/11646957/45746008-139bdd00-bbfa-11e8-83eb-e5a03d090f16.png">

bcmn commented:
UX reviewed this with Ngan Thi, noticed that whilst everything else seems correct, the background-colour was intended to have a 75% opacity. Currently the background-colour is entirely opaque.

bcmn committed:
Adds transparency to copyright background, ensures hidden accessibility full stop doesn't show against it

bcmn committed:
Updates Cypress tests to reflect new copyright background colour

bcmn committed:
Merge branch 'latest' of github.com:bbc/simorgh into integrate_copyright_element

sareh commented:
@bcmn Looks great.  👍 
I'll explain what I did when reviewing this change to the transparency. 

I ran the lighthouse tests against this branch `npm run test:lighthouse`. This runs against article http://localhost:7080/news/articles/c0000000027o It had an accessibility score of 100 still, so passes colour contrast.  
<img width="380" alt="screen shot 2018-09-20 at 08 52 45" src="https://user-images.githubusercontent.com/3028997/45804128-33430c00-bcb3-11e8-8a8c-beb896900f25.png">

I noted that this might have been because of the particular images in that article, so I ran the dev build `npm run dev`, visited that page http://localhost:7080/news/articles/c0000000027o and then removed the src for an image, so there was no background. 
<img width="745" alt="screen shot 2018-09-20 at 08 53 46" src="https://user-images.githubusercontent.com/3028997/45804619-89fd1580-bcb4-11e8-8dfa-ad2fb58dd701.png">


There is a very useful Color Contrast picker in Chrome Canary that now has in-built colour contrast guidance. (How to use tutorial is here https://remysharp.com/2017/08/17/contrast-ratio-in-devtools) 
It has a color contrast of 6.98, which more than meets our requirement (4.5) (https://bbc-news.github.io/accessibility-news-and-you/accessibility-news-and-testers)

<img width="1080" alt="screen shot showing a curve for where there's contrast" src="https://user-images.githubusercontent.com/3028997/45804558-50c4a580-bcb4-11e8-8158-b49b7098be78.png">

jtart reviewed (dismissed):
Looks great!

bcmn committed:
Changes horizontal padding on copyright element to half GEL spacing, and tests

bcmn committed:
Makes copyright padding definition more concise

bcmn commented:
Updated screenshots...

Third-party copyright, no caption:
<img width="790" alt="screen shot 2018-09-20 at 11 15 07" src="https://user-images.githubusercontent.com/11646957/45812149-dfdab900-bcc6-11e8-9dcb-310b928f36f9.png">

Third-party copyright, with caption:
<img width="775" alt="screen shot 2018-09-20 at 11 15 41" src="https://user-images.githubusercontent.com/11646957/45812163-e79a5d80-bcc6-11e8-9800-e91920f344e1.png">

First-party copyright, no caption:
<img width="798" alt="screen shot 2018-09-20 at 11 15 51" src="https://user-images.githubusercontent.com/11646957/45812183-f97c0080-bcc6-11e8-80d9-79f0ffdaa1eb.png">

Commit:
Merge branch 'latest' into integrate_copyright_element

bcmn commented:
Chatting with @nataliesmart, we've noticed that the invisible full stop to ensure a suitable pause when using AT offsets the text in a way that impacts the intended design. Once noticed, it's very obvious. I think we were curious if there was another way to handle this.

Personally, now we've got the visually hidden text for "image source", I wonder if we even need to continue worrying about the original impetus for placing the full stop in there? As far as I can recall, it originated out of wanting to avoid using VHT when it didn't seem 100% necessary, which is a boat that has now sailed. Could/should I just add the entirety of the copyright holder to the VHT, with punctuation to ensure a suitable pause afterwards?

⛵️

pjlee11 commented:
@bcmn as spoke in person I think we should keep the VHT seperate to the visible copyright element as it is becoming to intricate. Therefore we should have VHT with the prefix text and a full stop to ensure the accessibility standard are met, and then set aria-hidden and focusable=false on the visual element

bcmn committed:
Stops usage of invisible full-stop for accessibility. Now has dedicated VHT tag, containing full image source information, and the visual copyright element is given the aria-hidden attr.

bcmn committed:
Extracts hidden text prefix to constant, to avoid linting ever causing an issue with spacing

sareh committed:
Merge branch 'latest' into integrate_copyright_element

sareh committed:
Update proptype to string

Add hiddenText constant with concatenated strings, to avoid erroneous generated HTML comment appearing between 'Image source, ' and the children.

Co-authored-by: Ben <ben.newcombe@bbc.co.uk>

bcmn committed:
Merge branch 'latest' of github.com:bbc/simorgh into integrate_copyright_element

pjlee11 reviewed (dismissed):
Great work, hopefully no more requirements will change before it's merged  🎉

Commit:
Merge branch 'latest' into integrate_copyright_element

greenc05 commented:
@bcmn please could we talk about the visually hidden copyright code change? This has some implications for assistive technology.

pjlee11 commented:
@greenc05 @bcmn can we agree now that the work, unless failing test, can be merged as is then we will address any concerns in a follow up PR. I ask this because the PR has been open and bouncing across the board for almost 2 weeks due to a lack of UX/Accessibility communication and we should look not to hold up merging work. Thanks :)

Update: To clarify I mean the communication between UX and accessibility

greenc05 commented:
Sure @pjlee11

bcmn committed:
Merge branch 'latest' of github.com:bbc/simorgh into integrate_copyright_element
